File: scripts/add-dark-mode.py
#!/usr/bin/env python3
"""
Add Tailwind dark mode variants to all TSX files in app/ and components/ directories.
This script safely adds dark: prefixes to color classes that don't already have them.
"""

import os
import re
from pathlib import Path

# Dark mode transformation rules
DARK_MODE_RULES = {
    # Text colors
    r'\btext-gray-900\b': 'text-gray-900 dark:text-gray-100',
    r'\btext-gray-800\b': 'text-gray-800 dark:text-gray-200',
    r'\btext-gray-700\b': 'text-gray-700 dark:text-gray-300',
    r'\btext-gray-600\b': 'text-gray-600 dark:text-gray-400',
    r'\btext-black\b': 'text-black dark:text-white',

    # Background colors
    r'\bbg-white\b': 'bg-white dark:bg-gray-900',
    r'\bbg-gray-50\b': 'bg-gray-50 dark:bg-gray-950',
    r'\bbg-gray-100\b': 'bg-gray-100 dark:bg-gray-800',
    r'\bbg-gray-200\b': 'bg-gray-200 dark:bg-gray-700',
    r'\bbg-gray-300\b': 'bg-gray-300 dark:bg-gray-600',
    r'\bbg-purple-100\b': 'bg-purple-100 dark:bg-purple-900/20',
    r'\bbg-indigo-100\b': 'bg-indigo-100 dark:bg-indigo-900/20',

    # Border colors
    r'\bborder-gray-200\b': 'border-gray-200 dark:border-gray-700',
    r'\bborder-gray-300\b': 'border-gray-300 dark:border-gray-600',
    r'\bborder-white\b': 'border-white dark:border-gray-700',
}


def has_dark_variant_nearby(text, match_start, match_end):
    """Check if a dark: variant already exists near this match."""
    # Look ahead in the same className for dark: variants
    search_window = text[match_start:min(match_end + 100, len(text))]
    return 'dark:' in search_window


def process_file(file_path):
    """Process a single file and add dark mode variants."""
    print(f"Processing: {file_path}")

    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()

        original_content = content
        modifications = 0

        # Apply each transformation rule
        for pattern, replacement in DARK_MODE_RULES.items():
            # Only replace if the match doesn't already have a dark: variant nearby
            def replace_if_no_dark(match):
                nonlocal modifications
                start, end = match.span()

                # Check if dark: variant already exists in the className
                # Look for className boundaries (quotes)
                class_start = content.rfind('className', 0, start)
                if class_start == -1:
                    return match.group(0)

                # Find the closing quote of this className
                quote_start = content.find('"', class_start)
                if quote_start == -1:
                    quote_start = content.find("'", class_start)
                if quote_start == -1:
                    return match.group(0)

                quote_char = content[quote_start]
                quote_end = content.find(quote_char, quote_start + 1)
                if quote_end == -1:
                    return match.group(0)

                # Check if dark: exists in this className attribute
                classname_content = content[quote_start:quote_end]
                if 'dark:' in classname_content:
                    # Already has dark variants, skip
                    return match.group(0)

                modifications += 1
                return replacement

            content = re.sub(pattern, replace_if_no_dark, content)

        # Only write if changes were made
        if content != original_content:
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(content)
            print(f"  [OK] Added {modifications} dark mode variants")
            return modifications
        else:
            print("  - No changes needed")
            return 0

    except Exception as e:
        print(f"  [ERROR] Error processing file: {e}")
        return 0


def main():
    """Main execution function."""
    print("Adding Tailwind dark mode variants to TSX files...")
    print("=" * 60)
    print()

    # Get the project root (assuming script is in scripts/ directory)
    project_root = Path(__file__).parent.parent

    # Find all TSX files in app/ and components/
    tsx_files = []
    for directory in ['app', 'components']:
        dir_path = project_root / directory
        if dir_path.exists():
            tsx_files.extend(dir_path.rglob('*.tsx'))

    print(f"Found {len(tsx_files)} TSX files to process\n")

    total_files = 0
    total_modifications = 0

    for file_path in tsx_files:
        mods = process_file(file_path)
        if mods > 0:
            total_files += 1
            total_modifications += mods

    print()
    print("=" * 60)
    print("Summary:")
    print(f"  Files modified: {total_files}")
    print(f"  Total dark variants added: {total_modifications}")
    print("=" * 60)


if __name__ == '__main__':
    main()

File: scripts/add-dark-mode.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const glob = require('glob');

/**
 * Dark mode transformation rules
 * Maps light mode classes to their dark mode variants
 */
const DARK_MODE_RULES = {
  // Text colors
  'text-gray-900': 'text-gray-900 dark:text-gray-100',
  'text-gray-800': 'text-gray-800 dark:text-gray-200',
  'text-gray-700': 'text-gray-700 dark:text-gray-300',
  'text-gray-600': 'text-gray-600 dark:text-gray-400',
  'text-gray-500': 'text-gray-500 dark:text-gray-500', // middle gray, no change
  'text-black': 'text-black dark:text-white',

  // Background colors
  'bg-white': 'bg-white dark:bg-gray-900',
  'bg-gray-50': 'bg-gray-50 dark:bg-gray-950',
  'bg-gray-100': 'bg-gray-100 dark:bg-gray-800',
  'bg-gray-200': 'bg-gray-200 dark:bg-gray-700',
  'bg-gray-300': 'bg-gray-300 dark:bg-gray-600',
  'bg-purple-100': 'bg-purple-100 dark:bg-purple-900/20',
  'bg-indigo-100': 'bg-indigo-100 dark:bg-indigo-900/20',

  // Border colors
  'border-gray-200': 'border-gray-200 dark:border-gray-700',
  'border-gray-300': 'border-gray-300 dark:border-gray-600',
  'border-white': 'border-white dark:border-gray-700',
};

/**
 * Check if a class already has a dark: variant
 */
function hasDarkVariant(classString, baseClass) {
  const darkClass = `dark:${baseClass.replace(/^(text|bg|border)-/, '$1-')}`;
  return classString.includes('dark:');
}

/**
 * Process a single className attribute
 */
function processClassName(match, prefix, quote, content, suffix) {
  let modified = false;
  let newContent = content;

  // Split by spaces to get individual classes
  const classes = content.split(/\s+/).filter(Boolean);
  const processedClasses = new Set();
  const result = [];

  for (const cls of classes) {
    // Skip if already processed or if it's already a dark: variant
    if (processedClasses.has(cls) || cls.startsWith('dark:')) {
      result.push(cls);
      continue;
    }

    // Check if this class has a dark mode rule
    if (DARK_MODE_RULES[cls]) {
      // Check if dark variant doesn't already exist in the string
      const darkPattern = `dark:${cls.split('-').slice(0, 2).join('-')}`;
      const hasDark = classes.some(c => c.startsWith(darkPattern));

      if (!hasDark) {
        result.push(DARK_MODE_RULES[cls]);
        modified = true;
      } else {
        result.push(cls);
      }
      processedClasses.add(cls);
    } else {
      result.push(cls);
    }
  }

  if (modified) {
    newContent = result.join(' ');
    return `${prefix}${quote}${newContent}${quote}${suffix}`;
  }

  return match;
}

/**
 * Process a file and add dark mode variants
 */
function processFile(filePath) {
  console.log(`Processing: ${filePath}`);

  let content = fs.readFileSync(filePath, 'utf8');
  let modifications = 0;

  // Match className attributes with various quote styles
  // Handles: className="...", className='...', className={`...`}
  const classNameRegex = /(className\s*=\s*)(['"`])([^'"`]*?)\2/g;

  const newContent = content.replace(classNameRegex, (match, prefix, quote, classContent) => {
    const result = processClassName(match, prefix, quote, classContent, '');
    if (result !== match) {
      modifications++;
    }
    return result;
  });

  if (modifications > 0) {
    fs.writeFileSync(filePath, newContent, 'utf8');
    console.log(`  ✓ Added ${modifications} dark mode variants`);
    return modifications;
  } else {
    console.log(`  - No changes needed`);
    return 0;
  }
}

/**
 * Main execution
 */
function main() {
  console.log('Adding Tailwind dark mode variants to TSX files...\n');

  // Find all TSX files in app/ and components/ directories
  const appFiles = glob.sync('app/**/*.tsx', { absolute: true });
  const componentFiles = glob.sync('components/**/*.tsx', { absolute: true });

  const allFiles = [...appFiles, ...componentFiles];
  console.log(`Found ${allFiles.length} TSX files to process\n`);

  let totalFiles = 0;
  let totalModifications = 0;

  allFiles.forEach(file => {
    const mods = processFile(file);
    if (mods > 0) {
      totalFiles++;
      totalModifications += mods;
    }
  });

  console.log(`\n==================================`);
  console.log(`Summary:`);
  console.log(`  Files modified: ${totalFiles}`);
  console.log(`  Total dark variants added: ${totalModifications}`);
  console.log(`==================================`);
}

// Run the script
main();
